### PR TITLE
Migrate the workflow format from DSL1 to DSL2

### PR DESCRIPTION
**Pull request by justinjj24**
feature/dsl2-v1 is a migration of dsl1 to dsl2 format of branch develop (0.5.1)

### CODE REVIEW

**Comment by justinjj24** [comment]
Summary:
1) Migrate `main.nf` from `DSL1` format to `DSL2`
2) Code improvement `main.nf`
3) created new process `mosdepth_datamash `to post process of `mosdepth`
4) Removed `bin/run_mosdepth.sh`

File: main.nf [hunk 1]
```diff
@@ -220,31 +209,66 @@ process multiqc {
...
+	
+// input channels
+
+biosample_id = params.biosample_id
+
+reference = channel.fromPath(params.reference, checkIfExists: true)
+    .map{ fa -> tuple(fa, fa + ".fai") }
+
+input_file = file ( params.bam_cram )
+index_type = input_file.getExtension()
```
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:244]: `.getExtension()` not worked with `path`

File: nextflow.config [hunk 2]
```diff
@@ -68,3 +72,13 @@ dag {
     enabled = true
     file = "${params.infodir}/dag.pdf"
 }
+
+manifest {
+  name = 'npm-sample-qc'
+  author = 'Justin Jeyakani, Maxime HEBRARD, Rodrigo Toro, Nicolas Bertin'
```
Comment (mhebrard) [inline nextflow.config:78]: Add Mar as author

File: main.nf [hunk 3]
```diff
@@ -220,31 +209,66 @@ process multiqc {
...
+----------------------------------------------------------------------
+WORKFLOW
+---------------------------------------------------------------------
+*/
+	
+// input channels
+
+biosample_id = params.biosample_id
+
+reference = channel.fromPath(params.reference, checkIfExists: true)
```
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:240]: Agree, the `checkIfExists` propably better to keep all the checks outside the workflow

File: nextflow.config (hunk 2 above)
Comment (justinjj24) [reply nextflow.config]: Included Mar in the author list

File: main.nf [hunk 4]
```diff
@@ -220,31 +209,66 @@ process multiqc {
...
+    .map{ fa -> tuple(fa, fa + ".fai") }
+
+input_file = file ( params.bam_cram )
+index_type = input_file.getExtension()
+
+if (index_type == "bam")
+    cbam = channel.fromPath(params.bam_cram, checkIfExists: true)
```
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:247]: Agree, the `checkIfExists` propably better to keep all the checks outside the workflow in the next version

File: main.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (rodrigotoro) [reply main.nf:244]: I tested it, this seems to work:

```nextflow
nextflow.enable.dsl=2

test_path = Channel.fromPath("test.txt")
mapped = test_path.map {it -> it.getExtension()}
mapped.view()
```

Agreed that using `path ( params.bam_cram )` as a replacement for `file ( params.bam_cram )` here doesn't work.

File: main.nf [hunk 5]
```diff
@@ -68,149 +77,129 @@ LAUNCH INFO
...
 
-process cram_bam_index {
+process samtools_stats {
+    tag "${biosample_id}"
+    publishDir "${params.publishdir}/samtools", mode: "copy"
 
     input:
-    file reference from ref_fa_ch_cram_bam_index
-    file bam_cram from input_ch_cram_bam
+    tuple val(biosample_id), file(cbam), file(idx), file(fa), file(fai)
```
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:91]: all the process input tuple either `file` or `path` (latest) can be used, I don't remember why I choose `file` over `path`

[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:91]: Both input index and fasta index required by one or more tools in the workflow. mosdepth expect the cram and fasta index file, not remember picard or samtools expects one of the index file or only when the input is cram.

[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:91]: Changed to `path`

File: main.nf [hunk 6]
```diff
@@ -68,149 +77,129 @@ LAUNCH INFO
...
+}
 
-    publishDir "${params.publishdir}/samtools", mode: "copy"
+process mosdepth {
+    tag "${biosample_id}"
+    publishDir "${params.publishdir}/mosdepth", mode: "copy"
 
     input:
-    file "*" from cram_bam_ch_samtools_stats
+    tuple val(biosample_id), file(cbam), file(idx), file(fa), file(fai)
```
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:108]: Changed to `path`

File: main.nf [hunk 7]
```diff
@@ -68,149 +77,129 @@ LAUNCH INFO
...
-
+process mosdepth_datamash {
+    tag "${biosample_id}"
     publishDir "${params.publishdir}/mosdepth", mode: "copy"
 
     input:
-    file reference from ref_fa_ch_mosdepth
-    file "*" from cram_bam_ch_mosdepth
-    file gap_regions from gap_regions_ch_mosdepth
+    tuple file(fa), file(fai)
```
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:127]: Changed to `path`

File: main.nf [hunk 8]
```diff
@@ -68,149 +77,129 @@ LAUNCH INFO
...
 
 process picard_collect_multiple_metrics {
-
+    tag "${biosample_id}"
     publishDir "${params.publishdir}/picard", mode: "copy"
 
     input:
-    file reference from ref_fa_ch_picard_collect_multiple_metrics
-    file "*" from cram_bam_ch_picard_collect_multiple_metrics
+    tuple val(biosample_id), file(cbam), file(idx), file(fa), file(fai)
```
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:172]: Changed to `path`

File: main.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:244]: params.bam_cram (bam_cram: s3://npm-grids/gnanakkan/unittest/NA12878-chr22.bam)
Doesn't work with `path` or `.fromPath `

File: main.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinjj24) [reply main.nf:240]: The .fromFilePairs() will try in the next release. Will discuss further..
 
- In the params.yml we are giving full path of the bam/cram only and no index - `bam_cram: s3://1000genomes-dragen-3.7.6/data/individuals/hg38-graph-based/NA12878/NA12878.bam`

- 1k genome sample folders contain more than one bam and bai files, so cannot use wildcard

- The index file is not given in any of the command, further splitting or map only the bam from the pair required


channel.fromFilePairs(params.bam_cram, params.bam_cram + "bai", checkIfExists: true)
Channel.fromFilePairs("$baseDir/data/alignment/*.{bam,bai}", checkIfExists:true)

[the remark this answers is not in the document]
Comment (rodrigotoro) [reply main.nf:240]: Given that we want to make changes to the way bam/cram inputs are handled, we will close this for now and address it later on.

File: main.nf (hunk 4 above)
[the remark this answers is not in the document]
Comment (rodrigotoro) [reply main.nf:247]: We will update this also once we address bam/cram input strategy.

**Review by rodrigotoro — APPROVED** [review]

**Pull request merged**